metadata/src/main/java/org/apache/kafka/metadata/MetadataRecordSerde.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.metadata;
+
+import org.apache.kafka.common.metadata.MetadataRecordType;
+import org.apache.kafka.common.protocol.ApiMessage;
+import org.apache.kafka.common.protocol.ObjectSerializationCache;
+import org.apache.kafka.common.protocol.Readable;
+import org.apache.kafka.common.protocol.Writable;
+import org.apache.kafka.common.utils.ByteUtils;
+import org.apache.kafka.raft.RecordSerde;
+
+public class MetadataRecordSerde implements RecordSerde<ApiMessageAndVersion> {
+
+    @Override
+    public ObjectSerializationCache newWriteContext() {
+        return new ObjectSerializationCache();
+    }
+
+    @Override
+    public int recordSize(ApiMessageAndVersion data, Object context) {
+        ObjectSerializationCache serializationCache = (ObjectSerializationCache) context;
```

Comment:
The type here leaks over into `RaftClient` and into `RaftManager`, which is really annoying. Having `RaftClient<RecordType>` seems reasonable. An alternative is to change the type to always be `ObjectSerializationCache` since there isn't actually any other use case.

Reply:
I pushed a commit to do this. Let me know if that addresses the concern.